packages/odoo/test/Adaptor.test.js

```diff
@@ -0,0 +1,98 @@
+import { expect } from 'chai';
+
+import {
+  create,
+  update,
+  read,
+  deleteRecord,
+  setMockClient,
+} from '../src/Adaptor.js';
+
+const configuration = {
+  baseUrl: 'https://test.odoo.com/odoo',
+  password: 'somepassword',
+  username: 'someusername',
+  database: 'somedatabase',
+};
+
+const state = { configuration };
+
+describe('read record', () => {
+  it('should read a record', async () => {
+    const mock = {
+      read: (model, id, fields) => {
+        expect(model).to.eql('product.product');
+        expect(id).to.eql([2]);
+        expect(fields).to.eql(['name']);
+
+        return [
+          {
+            id: 2,
+            name: 'Saas Product',
+          },
+        ];
+      },
+    };
+    setMockClient(mock);
+
+    const { data } = await read('product.product', [2], ['name'])(state);
+    expect(data[0].id).to.eql(2);
+    expect(data[0].name).to.eql('Saas Product');
+  });
+});
+
+describe('create', () => {
+  it('should create a record', async () => {
+    const mock = {
+      create: (model, data, options) => {
+        expect(model).to.eql('res.partner');
+        expect(data).to.eql({ name: 'Jane Doe' });
+        expect(options).to.eql(23);
+        return 15;
+      },
+    };
+    setMockClient(mock);
+
+    const { data } = await create(
+      'res.partner',
+      { name: 'Jane Doe' },
+      { externalId: 23 }
+    )(state);
+    expect(data).to.eql(15);
+  });
+});
+
+describe('update', () => {
+  it('should update a record', async () => {
+    const mock = {
+      update: (model, id, data) => {
+        expect(model).to.eql('product.product');
+        expect(id).to.eql(4);
+        expect(data).to.eql({ name: 'Testing Product' });
+        return true;
+      },
+    };
+    setMockClient(mock);
+
+    const { data } = await update('product.product', 4, {
+      name: 'Testing Product',
+    })(state);
+    expect(data).to.eql(true);
+  });
+});
+
+describe('delete', () => {
+  it('should delete a record', async () => {
```

Comment:
What happens if we try to delete  a record that doesn't exist? Do we throw an exception? Should we?

Reply:
If we try to delete a record that does not exist, it does not throw an error. Instead, it just returns `true` whether the record is there or not

Reply:
Maybe if the client returns `false`, we should throw?

It can be hard to know whether this is the right behaviour. It's probably fine if you ask to delete an item that's been deleted.

One way we often fix this is to include an `options` object and pass `throwOnError: boolean` to control the behaviour. That feels good here but in the case of `delete` I think we should default to `false`.

For `update`, on the other hand, if the update failed I think that's probably an error because the database will not be in the state we asked for it to be. So we'd default `throwOnErrow` to `true` there.

Ok, this is a 1.0 release. Let's be brave and opinionated. `delete` should not throw, but `update` should. Don't add any options at this stage.

Reply:
If we can `create` should also throw if the requested thing wasn't created. I don't know if the client gives us feedback on that. Don't spend a lot of time on it but maybe take a look? :)

Reply:
@josephjclark  `update` throws an error automatically if the record is not found. I have added a test for the error

Reply:
@josephjclark  `update` throws an error automatically if the record is not found. I have added a test for the error